File: packages/php/src/Seguranca/Ataques/AtaqueAproximacaoLinear.php
<?php

namespace Application\Seguranca\Ataques;

use Application\Seguranca\AlvoCriptografico;
use Application\Seguranca\AtaqueInterface;
use Application\Seguranca\ResultadoAtaque;
use Application\Seguranca\SkipAtaqueException;

/**
 * Aproximação linear (criptoanálise de Matsui): procura por correlação entre
 * bits de entrada (IV) e bits de saída do keystream. Uma cifra ideal não tem
 * aproximação linear com viés relevante; bias alto é uma pista explorável.
 */
class AtaqueAproximacaoLinear implements AtaqueInterface
{
    public function __construct(
        private int $amostras = 200,
        private int $tamanhoBloco = 32,
        private float $limiteBias = 0.3,
    ) {
    }

    public function nome(): string
    {
        return 'Aproximação linear (viés de Walsh)';
    }

    public function executar(AlvoCriptografico $alvo): ResultadoAtaque
    {
        $tamIv = $alvo->tamanhoIv();
        $bitsEntrada = $tamIv * 8;
        $bitsSaida = $this->tamanhoBloco * 8;
        $tamChave = strlen($alvo->chaveDeTeste());

        $primeiro = $alvo->gerarKeystreamBruto(random_bytes($tamChave), random_bytes($tamIv), 'enc', $this->tamanhoBloco);
        if ($primeiro === null) {
            throw new SkipAtaqueException('Alvo não expõe gerarKeystreamBruto.');
        }

        $entradas = [];
        $saidas = [];

        for ($s = 0; $s < $this->amostras; $s++) {
            $chave = random_bytes($tamChave);
            $iv = random_bytes($tamIv);
            $ks = $alvo->gerarKeystreamBruto($chave, $iv, 'enc', $this->tamanhoBloco);

            $inBits = [];
            for ($j = 0; $j < $bitsEntrada; $j++) {
                $inBits[$j] = (ord($iv[$j >> 3]) >> (7 - ($j & 7))) & 1;
            }
            $outBits = [];
            for ($j = 0; $j < $bitsSaida; $j++) {
                $outBits[$j] = (ord($ks[$j >> 3]) >> (7 - ($j & 7))) & 1;
            }
            $entradas[] = $inBits;
            $saidas[] = $outBits;
        }

        $maiorBias = 0.0;
        $par = [-1, -1];
        for ($i = 0; $i < $bitsEntrada; $i++) {
            for ($j = 0; $j < $bitsSaida; $j++) {
                $iguais = 0;
                for ($s = 0; $s < $this->amostras; $s++) {
                    if ($entradas[$s][$i] === $saidas[$s][$j]) {
                        $iguais++;
                    }
                }
                $bias = abs($iguais / $this->amostras - 0.5);
                if ($bias > $maiorBias) {
                    $maiorBias = $bias;
                    $par = [$i, $j];
                }
            }
        }

        $vulneravel = $maiorBias > $this->limiteBias;

        return new ResultadoAtaque(
            $this->nome(),
            vulneravel: $vulneravel,
            severidade: $vulneravel ? 'alta' : 'info',
            detalhes: sprintf(
                'maior viés |p-0.5|=%.4f na aproximação IV[bit %d] -> saída[bit %d] (limite %s); %d amostras',
                $maiorBias,
                $par[0],
                $par[1],
                $this->limiteBias,
                $this->amostras,
            ),
            dados: ['maior_bias' => $maiorBias, 'par' => $par],
        );
    }
}

File: packages/julia/src/Seguranca/Ataques/AtaqueAproximacaoLinear.jl
"""
Aproximação linear (criptoanálise de Matsui): procura por correlação entre bits
de entrada (IV) e bits de saída do keystream. Uma cifra ideal não tem
aproximação linear com viés relevante; bias alto é uma pista explorável.
"""
struct AtaqueAproximacaoLinear <: Ataque
    amostras::Int
    tamanho_bloco::Int
    limite_bias::Float64
end

AtaqueAproximacaoLinear(;
    amostras::Integer = 200,
    tamanho_bloco::Integer = 32,
    limite_bias::Real = 0.3,
) = AtaqueAproximacaoLinear(Int(amostras), Int(tamanho_bloco), Float64(limite_bias))

nome(::AtaqueAproximacaoLinear) = "Aproximação linear (viés de Walsh)"

function executar(a::AtaqueAproximacaoLinear, alvo::AlvoCriptografico)
    tam_iv = tamanho_iv(alvo)
    bits_entrada = tam_iv * 8
    bits_saida = a.tamanho_bloco * 8
    tam_chave = length(codeunits(chave_de_teste(alvo)))

    entradas = zeros(Int, a.amostras, bits_entrada)
    saidas = zeros(Int, a.amostras, bits_saida)

    for s in 1:(a.amostras)
        chave = random_bytes(tam_chave)
        iv = random_bytes(tam_iv)
        ks = gerar_keystream_bruto(alvo, chave, iv, "enc", a.tamanho_bloco)

        for j in 0:(bits_entrada-1)
            entradas[s, j+1] = (iv[div(j, 8)+1] >> (7 - (j & 7))) & 0x01
        end
        for j in 0:(bits_saida-1)
            saidas[s, j+1] = (ks[div(j, 8)+1] >> (7 - (j & 7))) & 0x01
        end
    end

    maior_bias = 0.0
    par = (-1, -1)
    for i in 0:(bits_entrada-1)
        for j in 0:(bits_saida-1)
            iguais = 0
            for s in 1:(a.amostras)
                if entradas[s, i+1] == saidas[s, j+1]
                    iguais += 1
                end
            end
            bias = abs(iguais / a.amostras - 0.5)
            if bias > maior_bias
                maior_bias = bias
                par = (i, j)
            end
        end
    end

    vulneravel = maior_bias > a.limite_bias

    return ResultadoAtaque(
        nome(a),
        vulneravel,
        vulneravel ? "alta" : "info",
        "maior viés |p-0.5|=$(fmt(maior_bias, 4)) na aproximação IV[bit $(par[1])] -> saída[bit $(par[2])] (limite $(a.limite_bias)); $(a.amostras) amostras",
        Dict{String,Any}("maior_bias" => maior_bias, "par" => [par[1], par[2]]),
    )
end
